src/runtime_src/tools/xclbinutil/XclBinUtilities.cxx

```diff
@@ -909,7 +910,44 @@ transformMemoryBankGroupingCollections(const std::vector<boost::property_tree::p
   createMemoryBankGroupEntries(possibleGroupConnections, groupTopology, groupConnectivity);
 }
   
-
+bool
+XclBinUtilities::checkAIEPartitionIPLayoutCompliance(XclBin & xclbin){
+  // Get AIE_PARTITION metadata
+  std::set<std::string> allDpuKernelIDs;
+  Section *pAIEPartition = xclbin.findSection(AIE_PARTITION);
+  std::string jsonFile = pAIEPartition->getPathAndName();
+  boost::property_tree::ptree pt;
+  boost::property_tree::read_json(jsonFile, pt);
+  const boost::property_tree::ptree& ptAIEPartition = pt.get_child("aie_partition");
+  std::vector<boost::property_tree::ptree> ptPDIs = XUtil::as_vector<boost::property_tree::ptree>(ptAIEPartition, "PDIs");
+  for (const auto& pdi : ptPDIs) {
+    std::vector<boost::property_tree::ptree> ptCDOs = XUtil::as_vector<boost::property_tree::ptree>(pdi, "cdo_groups");
+    for (const auto& cdo_group : ptCDOs) {
+      std::vector<std::string> dpuKernelIDs = XUtil::as_vector_simple<std::string>(cdo_group, "dpu_kernel_ids");
+      allDpuKernelIDs.insert(dpuKernelIDs.begin(), dpuKernelIDs.end());
+    }
+  }
+                                             
+  // Get IP_LAYOUT metadata
+  Section *pIPLayout = xclbin.findSection(IP_LAYOUT);
+  boost::property_tree::ptree ptIPLayout;
+  pIPLayout->getPayload(ptIPLayout);
+  boost::property_tree::ptree ptIPlayout = ptIPLayout.get_child("ip_layout");
+  boost::property_tree::ptree ipDatas = ptIPlayout.get_child("m_ip_data");
```

Comment:
I thought we should use XUtil::as_vector here since m_ip_data is an array

Reply:
I am iterating over all the elements in ipDatas below in line 938 which is equivalent to your approach of using XUtil::as_vector (refer def of as_vector). But would need to iterate this vector to check the elements in this vector.